### PR TITLE
Allow AT Links to be created in the context of an Analysis Request

### PR DESCRIPTION
**Pull request by witekdev**
## Description of the issue/feature this PR addresses

This is a new feature that allows for Links to be created in the context of an Analysis Report.

The usecase:
When generating an ARReport sourced from multiple source ARs, currently an ARReport object (and csv and pdf data) is created in each of these source ARs objects. To prevent the duplication and to have only one canonical location where the report pdf and csv data are stored, only one ARReport object is created in the context of the first AR. The remaining ARs that are part of this multiply sourced ARReport, contain Links that point to this canonical ARReport object stored in the 1st AR.

## Current behavior before PR

Creating Links in the context of an AR is not allowed. 

## Desired behavior after PR is merged

Creating Links in the context of an AR is allowed.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/profiles/default/metadata.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <metadata>
-  <version>1.2.1</version>
+  <version>1.2.2</version>
```
Comment (xispa) [inline bika/lims/profiles/default/metadata.xml]: No need to change the version. We increase the version each time we publish a new release (the last published release is 1.2.0 and the next one to be released is 1.2.1). See https://github.com/senaite/senaite.core/releases

File: bika/lims/upgrade/configure.zcml [hunk 2]
```diff
@@ -101,4 +101,12 @@
...
        profile="bika.lims:default"/>
 
+ <genericsetup:upgradeDepends
+       description="Allow Link types in the context of an AR"
+       title="Upgrade to SENAITE.CORE 1.2.2"
+       source="1.2.1"
+       destination="1.2.2"
+       import_steps="typeinfo"
+       profile="bika.lims:default"/>
+
```
Comment (xispa) [inline bika/lims/upgrade/configure.zcml]: No need to add a new upgrade step. We do this each time we publish a new release. If your changes require some code to be included in an upgrade step, include the code in the upgrade step that matches with the current unreleased version.

File: setup.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 from setuptools import setup, find_packages
 
-version = '1.2.1'
+version = '1.2.2'
```
Comment (xispa) [inline setup.py]: No need to change version.

**Comment by witekdev** [comment]
Hi @xispa,
thanks for the quick reply and review!

With this PR we were not intending for the entire use case to go into core. We were intending for it to be placed in an add-on product. However if this PR (once cleaned up) could make it into core, it would simplify this add-on significantly. The other alternative for us is to use our own customized ARReport CT with Links added in the add-on, or manually adjust the Senaite instance that uses our add-on to allow for this (not best practice).

Relevant use case code and branch:
https://github.com/witekdev/als.lims/commit/de4b82c913f3d04fcf7611261d18a48514eb246d
https://github.com/witekdev/als.lims/commits/AN-99-published-COA-downloads-intelligent-titles

ping @mikejmets

File: bika/lims/profiles/default/metadata.xml (hunk 1 above)
Comment (witekdev) [reply bika/lims/profiles/default/metadata.xml]: :+1:

File: bika/lims/upgrade/configure.zcml (hunk 2 above)
Comment (witekdev) [reply bika/lims/upgrade/configure.zcml]: In this case there is no code change required, just to rerun portal_setup typeinfo.

File: setup.py (hunk 3 above)
Comment (witekdev) [reply setup.py]: :+1:

**Comment by witekdev** [comment]
@xispa if we included the complete functionality to support ARReport Links (as per usecase described) in this PR, would consider it for inclusion in senaite.core? Thanks!

**Comment by ramonski** [comment]
Hi @witekdev, thanks for your PR which is technically absolutely correct and even tested 👍 

However, what I don't like is the approach to use AT Link Objects as a reference holder between ARs and ARReports.

This link can break too easy (by deleting the link object), and then you don't know anymore which Report belongs to which AR.
Also the links does neither follow the nomenclature of the IDServer nor have any specific workflow attached.

I actually don't see the ability to add AT Links below ARs as a generic feature within the core, but would rather suggest to keep the code together with the rest of the functionality that implements that use-case.

**Comment by witekdev** [comment]
Thanks for the quick reply and feedback @ramonski 
Sounds good, we will figure a way to keep this functionality in an add-on.

**Pull request closed without merging**